polkadot/runtime/parachains/src/mock.rs

```diff
@@ -388,6 +388,121 @@ impl ValidatorSetWithIdentification<AccountId> for MockValidatorSet {
 	type IdentificationOf = FoolIdentificationOf;
 }
 
+pub mod mock_assigner {
+	use super::*;
+	pub use pallet::*;
+
+	#[frame_support::pallet]
+	pub mod pallet {
+		use super::*;
+
+		#[pallet::pallet]
+		#[pallet::without_storage_info]
+		pub struct Pallet<T>(_);
+
+		#[pallet::config]
+		pub trait Config: frame_system::Config + configuration::Config + paras::Config {}
+
+		#[pallet::storage]
+		pub(super) type MockAssignerQueue<T: Config> =
+			StorageValue<_, VecDeque<V0Assignment>, ValueQuery>;
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn add_on_demand_order(
+			order: V0Assignment,
+			location: QueuePushDirection,
+		) -> Result<(), DispatchError> {
+			MockAssignerQueue::<T>::try_mutate(|queue| {
+				// Don't need to worry about max assignment queue size in scheduler tests
+				match location {
+					QueuePushDirection::Back => queue.push_back(order),
+					QueuePushDirection::Front => queue.push_front(order),
+				};
+				Ok(())
+			})
+		}
+
+		fn is_legacy_core(core_idx: &CoreIndex) -> bool {
+			let parachain_cores = <paras::Pallet<Test>>::parachains().len() as u32;
+			(0..parachain_cores).contains(&core_idx.0)
+		}
+	}
+
+	impl<T: Config> AssignmentProvider<BlockNumber> for Pallet<T> {
+		type AssignmentType = V0Assignment;
+		type OldAssignmentType = V0Assignment;
+		// Format has not changed for parachains, therefore still version 0.
+		const ASSIGNMENT_STORAGE_VERSION: AssignmentVersion = AssignmentVersion::new(0);
+
+		fn migrate_old_to_current(
+			old: Self::OldAssignmentType,
+			_core: CoreIndex,
+		) -> Self::AssignmentType {
+			old
+		}
+
+		// This can matter, even in tests. Using condensed form of calculation from actual assigner.
+		fn session_core_count() -> u32 {
+			let config = <configuration::Pallet<Test>>::config();
+			(<paras::Pallet<Test>>::parachains().len() as u32)
+				.saturating_add(config.on_demand_cores)
+		}
+
+		// Create a new assignment to pass if core belongs to bulk buyer, else pop
+		// on demand assignment from front of queue.
+		fn pop_assignment_for_core(core_idx: CoreIndex) -> Option<Self::AssignmentType> {
+			if Self::is_legacy_core(&core_idx) {
+				<paras::Pallet<T>>::parachains()
+					.get(core_idx.0 as usize)
+					.copied()
+					.map(|para_id| V0Assignment::new(para_id))
+			} else {
+				let mut queue: VecDeque<V0Assignment> = MockAssignerQueue::<T>::get();
+				let front = queue.pop_front().map(|assignment| assignment);
+				// Write changes to storage.
+				MockAssignerQueue::<T>::set(queue);
+				front
+			}
+		}
+
+		// We don't care about core affinity in the test assigner
+		fn report_processed(_assignment: Self::AssignmentType) {}
+
+		fn push_back_assignment(assignment: Self::AssignmentType) {
+			if <paras::Pallet<T>>::is_parachain(assignment.para_id()) {
+			} else {
```

Comment:
Why are we doing this?

Reply:
For assignments corresponding to parachains we act as if we are in a real `ParachainsAssigner`, where there is no order queue to push back onto. 

This keeps bulk parachain assignments that are pushed back from accidentally adding to the on-demand queue.

Reply:
I'll add more comments describing what's going on here and throughout the mock assigner.

Reply:
But why? You seem to be basically reproducing the real functionality .... why have the mock in the first place then?

Reply:
Perhaps I misinterpreted this:

“We could and likely should use a test config that uses a mock test assignment provider, so we don't need to expose internals of the used assignment provider here”

I took this to mean, “Let’s construct a mock assignment provider with minimum necessary complexity to do the following.
1. Allow all scheduler tests to compile and pass without re-writing them.
2. Avoid exposing any internals of actual assigners to those tests.

It seems there’s a third criterion I’m not sure can be met without giving up the re-writing part of the first:
3. Avoid replicating any large portion of the existing assigner logic.

It seems like you’re not pleased with the way this mock assigner idea is turning out. Should it be scrapped?

Reply:
No, 3 is not directly the requirement. The requirement is, that if the interna of the assignment provider matter, then the scheduler tests are broken and should be fixed. Any assignment provider internal functionality that is tested there, should be tested in the assignment provider itself.